src/crimson/os/seastore/segment_cleaner.cc

```diff
@@ -306,22 +332,58 @@ SegmentCleaner::rewrite_dirty_ret SegmentCleaner::rewrite_dirty(
   Transaction &t,
   journal_seq_t limit)
 {
-  LOG_PREFIX(SegmentCleaner::rewrite_dirty);
   return ecb->get_next_dirty_extents(
     t,
     limit,
     config.journal_rewrite_per_cycle
   ).si_then([=, &t](auto dirty_list) {
+    LOG_PREFIX(SegmentCleaner::rewrite_dirty);
     DEBUGT("rewrite {} dirty extents", t, dirty_list.size());
     return seastar::do_with(
       std::move(dirty_list),
-      [FNAME, this, &t](auto &dirty_list) {
-	return trans_intr::do_for_each(
-	  dirty_list,
-	  [FNAME, this, &t](auto &e) {
+      [this, &t, limit](auto &dirty_list) {
+	journal_seq_t seq_to_rm = dirty_list.empty()
+	  ? limit
+	  : dirty_list.back()->get_dirty_from();
+	return backref_manager.batch_insert_from_cache(
+	  t,
+	  seq_to_rm,
+	  config.journal_rewrite_per_cycle
+	).si_then([this, &t, &dirty_list, seq_to_rm](auto inserted_to) {
+	  if (inserted_to != seq_to_rm && inserted_to != JOURNAL_SEQ_NULL) {
```

Comment:
`inserted_to < seq_to_rm` means that the amount of fresh backref extents has exceeded `config_t::journal_rewrite_per_cycle`, I wanted to avoid creating a transaction that's too large.

Reply:
Thanks! I just realized that `config.journal_rewrite_per_cycle` is passed to both `get_next_dirty_extents()` and `batch_insert_from_cache()` to control the CLEANER_TRIM transaction size. But it seems that definition of `journal_rewrite_per_cycle` has become vague because the actual CLEANER_TRIM size can be as much as 2 * `journal_rewrite_per_cycle`.

Further questions @xxhdx1985126 :
* Is there a requirement that dirty-tail must equal-or-less than the alloc-tail, or in other words, can these two tails be trimmed independently?
* Can we split the config to control rewrite_dirty_bytes_per_cycle and rewrite_alloc_bytes_per_cycle seperately?

Reply:
> * Is there a requirement that dirty-tail must equal-or-less than the alloc-tail, or in other words, can these two tails be trimmed independently?

Yes, I'm trying to break that invariant now.

> * Can we split the config to control rewrite_dirty_bytes_per_cycle and rewrite_alloc_bytes_per_cycle seperately?

yes, will do